user_sync/connector/directory_ldap.py

```diff
@@ -132,35 +144,52 @@ def load_users_and_groups(self, groups, extended_attributes, all_users):
-            if not user_subfilter.startswith('('):
-                user_subfilter = six.text_type('(') + user_subfilter + six.text_type(')')
-            group_user_filter = six.text_type('(&') + group_member_subfilter + user_subfilter + six.text_type(')')
             group_users = 0
-            try:
-                for user_dn, user in self.iter_users(group_user_filter, extended_attributes):
-                    user['groups'].append(group)
-                    group_users += 1
-                self.logger.debug('Count of users in group "%s": %d', group, group_users)
-            except Exception as e:
-                raise AssertionException('Unexpected LDAP failure reading group members: %s' % e)
+            if not two_steps_lookup:
+                group_dn = self.find_group_dn(group)
+                if not group_dn:
+                    self.logger.warning("No group found for: %s", group)
+                    continue
+                group_member_subfilter = self.format_ldap_query_string(group_member_filter_format, group_dn=group_dn)
+                if not group_member_subfilter.startswith('('):
+                    group_member_subfilter = six.text_type('(') + group_member_subfilter + six.text_type(')')
+                user_subfilter = all_users_filter
+                if not user_subfilter.startswith('('):
+                    user_subfilter = six.text_type('(') + user_subfilter + six.text_type(')')
+                group_user_filter = six.text_type('(&') + group_member_subfilter + user_subfilter + six.text_type(')')
+                try:
+                    for user_dn, user in self.iter_users(group_user_filter, extended_attributes):
+                        user['groups'].append(group)
+                        group_users += 1
+                except Exception as e:
+                    raise AssertionException('Unexpected LDAP failure reading group members: %s' % e)
+            else:
+                try:
+                    members = self.find_group_member_dn(group, two_steps_group_member_attribute_name)
+                    for member in members:
+                        user = all_users_records.get(member)
+                        if user:
+                            user['groups'].append(group)
+                            group_users += 1
+                except Exception as e:
+                    raise AssertionException('Unexpected LDAP failure reading group members: %s' % e)
+            self.logger.debug('Count of users in group "%s": %d', group, group_users)
 
         # if all users are requested, do an additional search for all of them
         if all_users:
             ungrouped_users = 0
             grouped_users = 0
             try:
-                for user_dn, user in self.iter_users(all_users_filter, extended_attributes):
+                for user_dn, user in all_users_records.iteritems():
                     if not user['groups']:
                         ungrouped_users += 1
```

Comment:
move the logic from below at line 203 right here, because you are doing the same test.  So first initialize a variable `grouped_user_records={}` before the loop and then in the loop do:
```python
if user['groups']:
    grouped_users += 1
    grouped_user_records[user_dn] = user
else:
    ungrouped_users += 1
```
Then after the loop do:
```python
if not all_users:
    self.user_by_dn = grouped_user_records
```